### ISSUE TITLE
PostgresDriver.preparePersistentValue for jsonb column should not call JSON.stringify if already a string

### ISSUE THREAD

**Issue by softwareplumber** [typeorm/typeorm#8128]
## Feature Description

I'd like to be able use a transformer to perform my own serialization/deserialization of json - for instance, using a parser which supports big integers. 

### The Problem

PostgresDriver.preparePersistentValue calls JSON.stringify even when the value given is already a string. What gets stored is therefore a quoted string containing the JSON one actually wants to store.

### The Solution

modify PostgresDriver.preparePersistentValue to pass through string values unchanged.

### Considered Alternatives

parametrize the JSON parser used in PostgresDriver so that (for example) json-bigint could be used.

### Additional Context

Large integers are pretty useful in the blockchain world - for instance IPLD's DAG-JSON codec uses many big integer fields.

### Relevant Database Driver(s)

| DB Type              | Relevant      |
| ---                  | ---           |
| `aurora-data-api`    | no            |
| `aurora-data-api-pg` | no            |
| `better-sqlite3`     | no            |
| `cockroachdb`        | no            |
| `cordova`            | no            |
| `expo`               | no            |
| `mongodb`            | no            |
| `mysql`              | no            |
| `nativescript`       | no            |
| `oracle`             | no            |
| `postgres`           | yes |
| `react-native`       | no            |
| `sap`                | no            |
| `sqlite`             | no            |
| `sqlite-abstract`    | no            |
| `sqljs`              | no            |
| `sqlserver`          | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✅ Yes, I have the time, and I know how to start.
 - ✖️ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by imnotjames** [comment]
I don't think we should be making that kind of change.  You'd break the case where someone wants to store a string that's not JSON in a JSON field.

**Comment by softwareplumber** [comment]
True. But:

1) So far as I can see the behavior I am proposing is the exact behavior 
of the node-postgres client: it passes a string straight through, and 
only calls JSON.stringify on an object.
2) There's a trivial workaround for the breaking case (just put quotes 
around the string), whereas at the moment there is no workaround at all 
for the limitations of JSON.stringify.

Set against this, 256 bit integers are a very important use case in the 
blockchain world.

Actually since node-postgres already seems to do what I want, maybe the 
simpler option would be to add a column option which makes 
PostgresDriver.preparePersistentValue pass any value it is given for a 
jsonb column straight through to node-postgres?

Regards
Jon.

------ Original Message ------
From: "James Ward" ***@***.***>
To: "typeorm/typeorm" ***@***.***>
Cc: "softwareplumber" ***@***.***>; "Author" 
***@***.***>
Sent: 8/28/2021 11:17:22 AM
Subject: Re: [typeorm/typeorm] PostgresDriver.preparePersistentValue for 
jsonb column should not call JSON.stringify if already a string (#8128)

**Comment by softwareplumber** [comment]
Okay, I think the correct thing to do is implement a behavior like for the hstore column type. 

This will maintain the exact same behavior as right now for all cases where an entity property with Object type is mapped to a json or jsonb field. Where the developer creates a *string* property and maps it to a jsonb field, it will map the json/jsonb directly out of postgres as a string. 

We also add a column option jsonType: 'object' | 'string' which is directly analogous to the hstoreType option for HSTORE columns.

Combining a jsonType option with a value transformer will allow a developer to do their own JSON parsing/stringifying using their choice of parser.

I have built and tested a prototype - all the existing unit tests pass.

Question: for new unit tests related to the new functionality, should I put them in an issue-specific test case, or expand the existing functional/jsonb.ts file?

**Comment by softwareplumber** [comment]
@imnotjames any chance you could take another look at this before the PR goes stale? The solution presented is much better than originally proposed.

**Comment by softwareplumber** [comment]
Hey, I know y'all are busy, but it's a little depressing watching the tumbleweed blow by...

**Comment by softwareplumber** [comment]
Still interested in this, if anyone is watching...

**Comment by softwareplumber** [comment]
Still hoping someone will take another look at this.
